src/de.hpi.swa.trufflesqueak/src/de/hpi/swa/trufflesqueak/nodes/primitives/impl/ControlPrimitives.java

```diff
@@ -669,7 +669,12 @@ protected static final AbstractPrimitiveNode createPrimitiveNode(final long prim
 
         protected static final Object primitiveWithArgs(final VirtualFrame frame, final Object receiver, final ArrayObject argumentArray,
                         final AbstractPrimitiveNode primitiveNode, final ArrayObjectToObjectArrayCopyNode toObjectArrayNode, final Node inlineTarget) {
-            return primitiveNode.executeWithArguments(frame, receiver, toObjectArrayNode.execute(inlineTarget, argumentArray));
+            try {
+                return primitiveNode.executeWithArguments(frame, receiver, toObjectArrayNode.execute(inlineTarget, argumentArray));
+            } catch (NegativeArraySizeException | ArrayIndexOutOfBoundsException oob) {
```

Comment:
I've looked into this and it seems you have fixed the symptoms but not the root cause. The problem appears to be that the nodes annotated with `@SqueakPrimitive(indices = 83)` don't check the argument count during dispatch.

`Interpreter>>#primitivePerform` does this:

```smalltalk
"Only test CompiledMethods for argument count - other objects will have to take their chances"
(self isOopCompiledMethod: newMethod)
	ifTrue: [self success: (self argumentCountOf: newMethod) = argumentCount].
```

I'd like us to fix the root cause instead but it's not trivial because the perform nodes reuse the `DispatchXNode` nodes and they don't check the argument count. One could derive the argument count from the selector, but that is not semantically identical to what OpenSmalltalkVM does.